doc/source/serve/advanced-guides/grpc-guide.md

```diff
@@ -0,0 +1,247 @@
+(serve-set-up-grpc-service)=
+# Set Up gRPC Service
+
+This section helps you understand how to:
+- build a user defined gRPC service and protobuf
+- start Serve with gRPC proxy
+- deploy gRPC applications
+- send gRPC requests to Serve deployments
+- health checks proxies
+- gRPC request metadata 
+- more examples
+- error handling
+
+
+(custom-serve-grpc-service)=
+## Define a gRPC Service
+To run a gRPC server it starts with first defining gRPC services, rpc methods, and 
+protobufs similar to the one below.
+
+```{literalinclude} ../doc_code/proxy_guide/user_defined_protos.proto
+:start-after: __begin_proto__
+:end-before: __end_proto__
+:language: proto
+```
+
+
+In this example, we created a file named `user_defined_protos.proto`. There are two
+gRPC services, `UserDefinedService` and `FruitService`. `UserDefinedService` has three
+RPC methods, `__call__`, `Multiplexing`, and `Streaming`. `FruitService` has one
+RPC method, `FruitStand`. Their corresponding input and output types are also defined
+specifically for each RPC method.
+
+Once the `.proto` services are defined, we can use `grpcio-tools` to compile python 
+code for those services. Example command looks like the following:
+```bash
+python -m grpc_tools.protoc -I=. --python_out=. --grpc_python_out=. ./user_defined_protos.proto
+```
+
+It will generate two files, `user_defined_protos_pb2.py` and 
+`user_defined_protos_pb2_grpc.py`.
+
+For more details on `grpcio-tools` see: [https://grpc.io/docs/languages/python/basics/#generating-client-and-server-code](https://grpc.io/docs/languages/python/basics/#generating-client-and-server-code)
+
+:::{note}
+Please make sure the generated files are in the same directory as where the Ray cluster
+is running so it can be importable by Serve when starting the proxies.
+:::
+
+(start-serve-with-grpc-proxy)=
+## Start Serve with gRPC Proxy
+[Serve start](https://docs.ray.io/en/releases-2.7.0/serve/api/index.html#serve-start) CLI 
+and [`ray.serve.start`](https://docs.ray.io/en/releases-2.7.0/serve/api/doc/ray.serve.start.html#ray.serve.start) API 
+both support starting Serve with gRPC proxy. There are two options related to Serve's 
+gRPC proxy, `grpc_port` and `grpc_servicer_functions`. `grpc_port` is the port for gRPC 
+proxies to listen. It defaults to 9000. `grpc_servicer_functions` is a list of import 
+paths for gRPC `add_servicer_to_server` functions to add to Serve’s gRPC proxy. It also 
+serves as the flag to determine whether to start gRPC server. Default empty list, 
+meaning no gRPC server will be started. 
+
+To start gRPC proxy with the CLI
```

Comment:
please also show how to use config file. you can maybe have 3 tabs (python, CLI, config file) here. look at the serve index page for how to do that in md.

Reply:
oh great idea, let me try to add tabs!